rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,869 @@
+- As a user, I want to know if a policy was successfully pulled, precompiled, and validated.
+- As a user, I want to monitor the status of a policy on a specific policy server replica so that I can identify and resolve issues with the policy.
+- As a user, I want the policy server to remain running with a previous version of the policy in case of issues with the latest policy.
+- As a user, I want to rollback to a previous policy version in case of issues with the latest policy.
+- As a user, I want to debug issues with the namespaced policies I own without involving the cluster administrator.
+
+# Detailed design
+
+[design]: #detailed-design
+
+> Note: In this RFC, the term `Policy` will be used to refer to any of the following resources: `AdmissionPolicy`, `ClusterAdmissionPolicy`, `AdmissionPolicyGroup`, or `ClusterAdmissionPolicyGroup`.
+
+## Concepts
+
+To implement the new policy lifecycle, we introduce the following concepts:
+
+### Hot reload
+
+The Policy Server will be able to load a new policy without requiring a restart.
+This will be achieved by refactoring the `EvaluationEnvironment` and adding methods to set, update, and remove policies.
+
+### The Policy Server acts as a controller
+
+Currently, the Policy Server is configured with a ConfigMap that contains all the policies to be loaded.
+When a new policy is created, the ConfigMap is updated, and the Policy Server is restarted to load the new policy.
+
+We want to change this behavior to allow the Policy Server to act as a controller that can load new policies without requiring a restart.
+
+#### PolicyRevision CRD
+
+We will introduce a new CRD called `PolicyRevision` to store the state of a policy in relation to the Policy Server replicas.
+`PolicyRevisions` are namespaced resources, and they will be created in the same namespace as the Kubewarden controller.
+
+The `PolicyRevision` will include fields to store the policy content and an `enabled` field to specify whether the policy should be served by the Policy Server.
+The `PolicyRevision` will include a `policyGeneration` field to store the generation of the associated `Policy`.
+This field will be set to the value of `metadata.generation` from the `Policy` CRD at the time the `PolicyRevision` is created.
+This approach allows the Policy Server to manage multiple generations of a policy at the same time, facilitating the gradual rollout of updates.
+
+This is similar to the `Deployment` and `ReplicaSet` relationship in Kubernetes, where the `Deployment` defines the desired state at a higher level of abstraction,
+and the `ReplicaSet` is responsible for managing the current state to align with this desired state.
+Additionally, `ReplicaSets` also store the backup of older `Deployment` revisions, enabling rollback to a previous state if needed.
+The `PolicyRevision` will also store the older policy to a maximum of `n` revisions, where `n` is a configurable value.
```

Comment:
I think the current design only accounts for 1 old revision, not `n` (or I don't understand the specifics on how the rollback and garbage collection happens).

Reply:
Resolving, I understand that we just have several PolicyRevisions, one per revision we want.